Use a single config file

victorlin opened the pull request:
## Description of proposed changes

This is an alternative to #101 and the preferred approach discussed at dev chat today.

<details>
<summary>
updated workflow diagram
</summary>

<img width="1400" height="930" alt="dag" src="https://github.com/user-attachments/assets/cb9f4565-655a-464b-83cd-a4592c8e1250" />

</details>

[Preview of changes for the NW-PaGe fork (mainly the last commit)](https://github.com/NW-PaGe/WNV/compare/main...nextstrain:WNV:victorlin/single-config-nw-page)

## Related issue(s)

- Closes #100

## Discussion threads

- [x] [Should `augur subsample` use the entire dump of Snakemake YAML and read from a config key path?](https://github.com/nextstrain/WNV/pull/102#discussion_r2310650152) yes
    - [x] [Is it ok to require Augur at runtime?](https://github.com/nextstrain/WNV/pull/102#discussion_r2311685873) no longer necessary
- [x] [Should the default behavior be to run all 3 Nextstrain builds?](https://github.com/nextstrain/WNV/pull/102#discussion_r2317316431) yes
    - [x] [How do we structure multi-build configs – build-first or rule-first?](https://github.com/nextstrain/WNV/pull/102#discussion_r2320205454) build-first for this workflow
    - [x] [Rename `builds_to_run` + `builds`?](https://github.com/nextstrain/WNV/pull/102#discussion_r2323244976) renamed to `builds` + `build_params`
- [x] [Move everything but `strain_id_field` and `inputs` to build-specific config?](https://github.com/nextstrain/WNV/pull/102#discussion_r2317255081) yes
- [x] [Add validation between `builds` and `builds_to_run`?](https://github.com/nextstrain/WNV/pull/102#discussion_r2317269400) yes
- [x] [Write config in a rule or as part of top-level workflow code?](https://github.com/nextstrain/WNV/pull/102#discussion_r2317320380) yes
    - [x] [Write config with timestamps in the name?](https://github.com/nextstrain/WNV/pull/102#discussion_r2319782950) yes
- [x] [Where is config written?](https://github.com/nextstrain/WNV/pull/102#discussion_r2317323204) Snakemake workdir
- [x] [Rename `--config-path`?](https://github.com/nextstrain/WNV/pull/102#discussion_r2317432900) renamed to `--config-section`

## Checklist

<!--
Make sure checks are successful at the bottom of the PR.

If applicable, add:
- any changes to existing tests
- any additional manual testing to confirm changes

Please add a note if you need help with adding tests.
-->

- [x] Checks pass
- [x] Update changelog
- [x] Pre-merge: drop 49dcc16502518302b3982bbb8ed6f288c2dc12dc (to be added in a new PR)
- [x] Pre-merge: rebase to remove augur subsample commits and replace "TBD" in changelog

<!-- 🙌 Thank you for contributing to Nextstrain! ✨ -->

victorlin commented on phylogenetic/rules/subsampling_configurable.smk:
```diff
@@ -15,58 +15,48 @@ OUTPUTS:
+    run:
+        import yaml
+        with open(output.config, 'w') as f:
+            yaml.dump(config["builds"][wildcards.build]["subsample"], f, sort_keys=False)
```
Re: 9ee5ce9bccb08a7338cf9281406f76b226201b6b

> (not pursued) Update 'augur subsample' to take a '--config-path' instead. That could get weird with the need to handle special characters.

I take this back, it's worth considering to simplify the Snakemake workflow. I'll try it out today.

victorlin replied:
Done:

https://github.com/nextstrain/WNV/blob/b77a4e4439d8df652a9abe79a0d49b8c052bd915/phylogenetic/rules/subsample.smk#L23-L52

victorlin replied:
Full context [from Slack](https://bedfordlab.slack.com/archives/C01LCTT7JNN/p1756504593956529):

- [--config-root](https://github.com/nextstrain/augur/commit/65413dc91333b8f7264344b819a940d2d0dd5053) has been replaced by [--config-path](https://github.com/nextstrain/augur/commit/5432d5757cb2b2fe107021f293485203f721c6f3)
- [filtering and subsampling guide](https://nextstrain--257.org.readthedocs.build/en/257/guides/bioinformatics/filtering-and-subsampling.html#within-snakemake-workflow-config) now recommends using a [format_config_path](https://nextstrain--1879.org.readthedocs.build/projects/augur/en/1879/api/public/augur.subsample.html) helper function
- wnv draft (this PR) implements the above recommendation

huddlej replied:
You may have discussed this elsewhere, but this pattern of importing an augur module from the top-level of a Snakefile will not work as expected when running the workflow with `snakemake --use-conda` like we* do from cluster environments.

>  from augur.subsample import format_config_path 
>  
> rule subsample: 

To make this work, the user will need to run Snakemake from an environment on the cluster with Augur available at runtime. This is an issue for other pathogen workflows, too (e.g., [seasonal flu](https://github.com/nextstrain/seasonal-flu/blob/cd564882ed63cf1b791f8ac0651be69cd7cdd468/Snakefile#L2-L3)).

(*I'm not sure how many people this "we" represents anymore!)

victorlin replied:
Thanks for flagging @huddlej, that usage had not crossed my mind. I think it's fine for a workflow to require an environment with Augur available at runtime, just like how it already requires Snakemake at runtime.

victorlin replied:
Actually, `augur subsample` could take a list of strings as the key path. This would remove the need for any special handling of characters and work nicely with Snakemake's string handling:

```
params:
    config_path = ["builds", w.build, "subsample"]
shell:
    """
    augur subsample --config-path {params.config_path:q}
```

I'll try it out today.

victorlin replied:
The list of strings worked out well.

https://github.com/nextstrain/WNV/blob/fb64a8e3d25b707ac809fa1f5e47a55dc55766ab/phylogenetic/rules/subsample.smk#L36-L51

victorlin committed:
Remove option for manual subsampling

This will not work well with multiple builds in a single workflow run.

victorlin committed:
Write run time config to a file

This is generally useful for debugging and will be used by augur
subsample.

Note that while this is under the 'rules' folder, it is not a rule.
Instead, it is top-level workflow code run before rules to ensure that
the file is written on every workflow run, including dry runs.

joverlee521 commented on phylogenetic/defaults/config.yaml:
```diff
@@ -11,22 +7,72 @@ inputs:
+      treetime_params: --coalescent opt --date-inference marginal --date-confidence --keep-polytomies --clock-rate 0.000273
+
+    export:
+      auspice_config: "defaults/lineage-2/auspice_config.json"
```
The split of nested build params vs top-level workflow params can be confusing here without digging into the workflow code, e.g. `builds.<build>.export.auspice_config` vs `export.description`. If at some point in the future we want per build descriptions, then we'd have to make workflow changes and breaking changes in the config to move it to build params.

If we're making this change, I suggest moving everything after `inputs` to per build params, even if that means we need to repeat some values.

victorlin replied:
Good point. I updated b57e1f901b138d022c34ad3a6d3233bf9edde23b to move everything after `inputs` to build-specific options.

Note the updated changelog entries for that commit:

https://github.com/nextstrain/WNV/blob/b57e1f901b138d022c34ad3a6d3233bf9edde23b/CHANGELOG.md#L7-L8

joverlee521 commented on phylogenetic/Snakefile:
```diff
@@ -8,7 +8,7 @@ workdir: workflow.current_basedir
 configfile: "defaults/config.yaml"
 
-builds = [build for build in config['builds'] if config['builds'][build] is not None]
+builds = config['builds_to_run']
```
_non-blocking_

If separating `builds` and `builds_to_run`, it'd be helpful to assert that `builds_to_run` is a subset of `builds` and write a helpful error message. The default Snakemake message is not super informative

```
$ nextstrain build phylogenetic/ --config builds_to_run='["foo"]'
Using workflow specific profile profiles/default for setting default command line arguments.
Assuming unrestricted shared filesystem usage.
host: 00b711764cb6
Building DAG of jobs...
InputFunctionException in rule export in file "/nextstrain/build/phylogenetic/rules/export.smk", line 29:
Error:
  KeyError: 'foo'
Wildcards:
  build=foo
Traceback:
  File "/nextstrain/build/phylogenetic/rules/export.smk", line 38, in <lambda> (rule export, line 29, /nextstrain/build/phylogenetic/rules/export.smk)
```

victorlin replied:
Thanks, I'll add some validation if the decision in https://github.com/nextstrain/WNV/pull/102#discussion_r2317316431 is to keep using `builds_to_run`.

victorlin replied:
Done:

https://github.com/nextstrain/WNV/blob/b0c8bec3524da7d8a694b0049e6af4d1bf591a2a/phylogenetic/Snakefile#L11-L27

This looks like:

```
ERROR: The following names in 'builds_to_run' are not defined in 'builds':

    foo

Available builds are:

    all-lineages
    lineage-1A
    lineage-2
```

victorlin commented on phylogenetic/defaults/config.yaml:
```diff
@@ -11,22 +7,77 @@ inputs:
+builds_to_run:
+  - all-lineages
+  - lineage-1A
+  - lineage-2
```
I think it's weird to have Nextstrain-specific config in defaults and the need to duplicate build names in `builds_to_run` for the sake of disabling those defaults. Another approach would be to move Nextstrain builds configuration to its own file: 157272f1965af489ae0bcd149fd20a7e04a00a1a

That approach seemed good to me initially, but it is naturally followed by changes such as 138e1e7086ee6fd1ce3abc80ad68e332a2c5a2bf and ce4c7d2af3ab92739054344df46c70c871cc9419 which I'm not so sure of (is everything in phylogenetic/defaults/config.yaml "Nextstrain-specific"?). It also goes against the [recommendation](https://github.com/nextstrain/pathogen-repo-guide/blob/f3c792c2a4d6ebec3a70c3d65b61258208789c67/phylogenetic/defaults/config.yaml#L1-L2) in pathogen-repo-guide:

> This configuration file should contain all required configuration parameters for the phylogenetic workflow to run to completion.

joverlee521 replied:
Personally, I'm a fan of the initial change to use `~` to disable default builds.

https://github.com/nextstrain/WNV/blob/0ec26ce8a83bf9157a29be0604145ae78e6c4960/phylogenetic/build-configs/ci/config.yaml#L10-L12

victorlin replied:
I think it's poor user experience – any time we add a new Nextstrain dataset under the default builds, users would have to update their config files to explicitly exclude it.

joverlee521 replied:
FWIW, having a list of what to run + a separate dict of config params is the general pattern that our multi-build configs follow (e.g. [avian-flu](https://github.com/nextstrain/avian-flu/blob/f1657094ebf9da39660a7f4ba6a5ddf51b6a9c3e/genome-focused/config.yaml#L4-L20), [mumps](https://github.com/nextstrain/mumps/blob/82bf417711d2246e333fd3f2d47a3888c85854ea/phylogenetic/defaults/config.yaml#L3), [seasonal-cov](https://github.com/nextstrain/seasonal-cov/blob/2ce3e7e51531f4bbaceb1985d1c9db61bab992bb/phylogenetic/defaults/config.yaml#L4-L8))

jameshadfield replied:
> FWIW, having a list of what to run + a separate dict of config params is the general pattern that our multi-build configs follow (e.g. [avian-flu](https://github.com/nextstrain/avian-flu/blob/f1657094ebf9da39660a7f4ba6a5ddf51b6a9c3e/genome-focused/config.yaml#L4-L20), [mumps](https://github.com/nextstrain/mumps/blob/82bf417711d2246e333fd3f2d47a3888c85854ea/phylogenetic/defaults/config.yaml#L3), [seasonal-cov](https://github.com/nextstrain/seasonal-cov/blob/2ce3e7e51531f4bbaceb1985d1c9db61bab992bb/phylogenetic/defaults/config.yaml#L4-L8))

The config structure of this PR is quite different from what we've done elsewhere. I don't think there are any other repos which have what are essentially single-build configs co-located like this (e.g. `builds.lineage-1A` is what I'm considering akin to a "single-build config"). Perhaps I've missed something here? @victorlin could you expand a little on what you see as the advantages of this approach?

victorlin replied:
Thanks for the links, I hadn't considered how this aligns with other pathogen repos. This PR follows the structure proposed in https://github.com/nextstrain/pathogen-repo-guide/issues/79.

I see no clear consensus on how to structure config options that differ between builds in a single config file. Each of the 3 examples uses its own approach:

1. "build-first" with config key paths as `<build>.<rule>.option`. This is what's proposed in [nextstrain/pathogen-repo-guide#79](https://github.com/nextstrain/pathogen-repo-guide/issues/79), used in seasonal-cov, and used in this PR.
2. "rule-first" with config key paths as `<rule>.<build>.option`. This is what's used in avian-flu and half of mumps.
3. "rule-only" with build templated within the value. This is what's used in the other half of mumps. Example: `filter.exclude: "{build}/exclude.txt"`

I think the rule-only approach should be replaced by either build-first or rule-first.

The decision between build-first and rule-first depends on whether it is more beneficial to compare config values across rules or builds. For example, in the avian-flu example, it's easy to see the difference in min length configs across segments, but it's difficult to figure out the exact config of the `h5n1-cattle-outbreak/ha/2y` build.

One more plus for each:
- The build-first approach makes it easier to make build-level modifications such as adding or deleting a build entirely.
- The rule-first approach allows wildcards in build parts (used in avian-flu), which isn't possible with the build-first approach (unless adding another templating layer such as CUE?).

jameshadfield replied:
My 2 cents (although I'm trying to step back from too many snakemake discussions).

When adding a new build (especially for testing out a scientific idea, which I find very common) I want a config design which allows me to do this cheaply. "Build-first" is not cheap in this sense, even if you utilise anchors. "Rule-first" _can_ be cheap if you have the concept of fallbacks / defaults. I don't understand "rule-only" in the context of parameter values, but for the example you give it's not cheap as you're having to provision an exclude file for your new build. I hope this goes some way to help understand why I designed avian-flu the way I did.

> but it's difficult to figure out the exact config of the h5n1-cattle-outbreak/ha/2y build

There's been various conversations that we should dump out the exact per-build (e.g. for "h5n1-cattle-outbreak/ha/2y") config at run time. Combined with a quick feeback loop like `snakemake results/h5n1-cattle-outbreak/ha/2y/config.yaml` it should help a lot.

victorlin replied:
> I don't understand "rule-only" in the context of parameter values

It's not valid, mumps uses "rule-first" in that case. Updated my previous comment to reflect that.

joverlee521 replied:
Sorry, didn't mean to derail this thread...just wanted to note that it's a relatively normal pattern to explicitly list which builds to run separately from the builds' config params.

I think decision on build-first vs rule-first will differ based on the pathogen. For something like flu, there's multiple levels of wildcards that define a build, so it's "cheaper" to start up a new build with the rule-first globbing syntax. For WNV, there's only a single level wildcard that define a build, so build-first is the less complex config that is easier to start with.

victorlin replied:
No worries, it was insightful to think about the different approaches and give them names. I'll plan to keep this PR as-is (build-first with `builds_to_run`).

jameshadfield replied:
> I'll plan to keep this PR as-is (build-first with builds_to_run).

Could you make `builds` the list of (string-value) builds to run, so that this aspect of "build-first" and "rule-first" config patterns is the same. That is

```
builds:
  - all-lineages
  - lineage-1A
  - lineage-2
```

I don't have a great suggestion for the key to store the actual config blocks under, `params`, `config`, `info` etc would all work but none are obviously the best word to use.

victorlin replied:
I also can't think of a good name for the other key, and for that reason I'm inclined to keep it as-is with `builds_to_run` + `builds`.

seasonal-cov uses `viruses` + individual top-level build keys, but I think it's good to put them under a dedicated key.

victorlin replied:
Going back to the first point in this thread:

> I think it's weird to have Nextstrain-specific config in defaults

This was discussed at dev chat today. The consensus is that defaults should aim to reproduce what's hosted on nextstrain.org, so having Nextstrain-specific config in defaults is preferred. I'll keep things as-is for this PR, summarized by the changelog entry:

https://github.com/nextstrain/WNV/blob/b0c8bec3524da7d8a694b0049e6af4d1bf591a2a/CHANGELOG.md#L9

joverlee521 replied:
> I also can't think of a good name for the other key, and for that reason I'm inclined to keep it as-is with builds_to_run + builds.

`builds` is shorter which is slightly nicer for CLI `--config` overrides...How about `builds` + `build_params`?

victorlin replied:
That sounds fine. I also realized that `builds` is a list in the current implementation, so keeping it as a list would be somewhat less of a breaking change.

I've combined 96bca923c751af243da1ede276c133018dde2c1e + bd0f5c26db0238b49940b5c0e8b6c16661ab6450 into one commit e676a85f14f85973200d3f17fc065611e02b4049.

joverlee521 replied:
(Just noticed that as of Snakemake 9.0.0, there is a `--replace-workflow-config` flag to fully replace the whole config without merging with the default config. Still nice to have the split of `builds` + `build_params`, but just wanted to flag it since it could be an alternative to disabling default builds)

victorlin replied:
Oh neat, I can see that being useful in the future.

joverlee521 commented on phylogenetic/rules/write_config.smk:
```diff
@@ -0,0 +1,16 @@
+
+path = "results/run_config.yaml"
+
+os.makedirs(os.path.dirname(path), exist_ok=True)
+
+with open(path, 'w') as f:
+    yaml.dump(config, f, sort_keys=False)
```
_non-blocking_

Since this is not a wrapped in a Snakemake rule, there's no indication in the workflow logs that this file is created. It'd be useful to add a print statement here that can be found in the logs

```suggestion
os.makedirs(os.path.dirname(path), exist_ok=True)

print(f"Writing the current run config to {path}")
with open(path, 'w') as f:
    yaml.dump(config, f, sort_keys=False)
```

Also, consider that if we don't wrap this in a Snakemake rule, we lose the built-in check for ambiguous rules that produce the same output file. We can potentially write a rule in the workflow that overwrite this file without realizing it.

victorlin replied:
I added the following message:

https://github.com/nextstrain/WNV/blob/4677d6abb04f3670c42b3c4a0b6792e04de73294/phylogenetic/rules/write_config.smk#L18

I'm torn between what I have now and adding this as a rule with an [always-touched file input](https://stackoverflow.com/a/36912085). Pros and cons:

|        | Pros                                             | Cons                                                                 |
|--------|--------------------------------------------------|----------------------------------------------------------------------|
| No rule| - runs with `--dry-run`                          | - needs an extra print statement<br>- can be overwritten by rules    |
| Rule   | - prevents other rules from overwriting          | - needs an always-touched file input<br>- does not run with `--dry-run` |

joverlee521 replied:
Ah, totally get the desire to produce the config file with `--dry-run`. 
Maybe prefix with a timestamp? `results/<timestamp>/run_config.yaml`?

victorlin replied:
`results/<timestamp>/run_config.yaml` would make the files slightly less accessible with intermediate directories containing only the single file. How about `results/run_config_<timestamp>.yaml`?

```
phylogenetic/results
├── all-lineages
├── lineage-1A
├── lineage-2
├── metadata_merged.tsv
├── run_config_2025-09-03_20:23:18.yaml
├── run_config_2025-09-03_20:23:21.yaml
├── run_config_2025-09-03_20:23:25.yaml
├── run_config_2025-09-03_20:23:28.yaml
├── run_config_2025-09-03_20:23:31.yaml
├── run_config_2025-09-03_20:23:33.yaml
├── sequences_merged.fasta
└── wa
```

Or we could keep `results/` tidy with `results/run_configs/<timestamp>.yaml`:

```
phylogenetic/results
├── all-lineages
├── lineage-1A
├── lineage-2
├── metadata_merged.tsv
├── run_configs
│   ├── 2025-09-03_20:27:55.yaml
│   ├── 2025-09-03_20:27:59.yaml
│   ├── 2025-09-03_20:28:07.yaml
│   ├── 2025-09-03_20:28:12.yaml
│   ├── 2025-09-03_20:28:18.yaml
│   └── 2025-09-03_20:28:20.yaml
├── sequences_merged.fasta
└── wa
```

joverlee521 replied:
Ah, yeah, keeping things tidy under `results/run_configs/` looks similar to snakemake logs
```
.snakemake/log/
├── 2025-07-11T234804.051801.snakemake.log
├── 2025-07-12T000059.269843.snakemake.log
├── 2025-07-29T230022.746369.snakemake.log
├── 2025-08-05T180716.665726.snakemake.log
├── 2025-08-05T180743.185104.snakemake.log
├── 2025-08-05T181254.063276.snakemake.log
├── 2025-08-05T181258.495974.snakemake.log
├── 2025-08-05T181327.652119.snakemake.log
├── 2025-08-05T181723.669536.snakemake.log
├── 2025-08-05T182014.222814.snakemake.log
├── 2025-08-05T182129.293584.snakemake.log
├── 2025-09-02T220334.626907.snakemake.log
└── 2025-09-02T221135.782401.snakemake.log
```

victorlin replied:
Done:

https://github.com/nextstrain/WNV/blob/65ad0f62ce50e6b8f1e0b43a4bf939c06ef05330/phylogenetic/rules/write_config.smk#L13-L14

I also updated the print statement to use stderr to avoid interfering with `--dag` outputs

https://github.com/nextstrain/WNV/blob/65ad0f62ce50e6b8f1e0b43a4bf939c06ef05330/phylogenetic/rules/write_config.smk#L21

and updated the note in the commit description to

> Note that while this is under the 'rules' folder, it is not a rule. Instead, it is top-level workflow code run before rules to ensure that the file is written on every workflow run, including dry runs.

victorlin replied:
Oh...

> Error: The path for one of the files in artifact is not valid: /phylogenetic/results/run_configs/2025-09-03_20:54:43.yaml. Contains the following character:  Colon :
> Invalid characters include:  Double quote ", Colon :, Less than <, Greater than >, Vertical bar |, Asterisk *, Question mark ?, Carriage return \r, Line feed \n

Renamed the file in 9b75182ab6c076696edf2ca57959ac040d86d884 to use the same format as Snakemake's log file. Not that it matters, but interesting to see that run config files are created slightly before log files:

```
2025-09-03T210400.622845.yaml
2025-09-03T210400.634809.snakemake.log
```

Also, the code is written to support local timezones, however it stays UTC when using the Docker runtime because host machine timezone is not passed into the container ([and doing so is not trivial](https://forums.docker.com/t/synchronize-timezone-from-host-to-container/39116)).

victorlin replied:
I'm reverting back to `results/run_config.yaml` in the `augur subsample` PR. Reasoning in e1d00afb8657e7182107ef446a64da453f33678c

joverlee521 commented on phylogenetic/rules/write_config.smk:
```diff
@@ -0,0 +1,16 @@
+import os
+import yaml
+
+path = "results/run_config.yaml"
```
_non-blocking_

I assume this is writing to the workdir? (Not relevant now since repo==workdir, but will be relevant when migrating to `nextstrain run`)

victorlin replied:
It's writing to Snakemake's workdir (`./phylogenetic/`), not ours (`./`).

https://github.com/nextstrain/WNV/blob/4677d6abb04f3670c42b3c4a0b6792e04de73294/phylogenetic/Snakefile#L5-L6

joverlee521 replied:
Ah okay. That statement will be removed when we add `nextstrain run` support so it should then output to whatever workdir that is set.

joverlee521 commented on phylogenetic/rules/subsample.smk:
```diff
@@ -15,56 +15,38 @@ OUTPUTS:
-            --exclude-all \
-            --include {input.subsampled_strains} \
+            --config {input.config} \
+            --config-path {params.config_path} \
```
_non-blocking_ 

`--config-path` makes me think it's a file path...thoughts on `--config-key`?

victorlin replied:
`--config-key` sounds good, I'll make the change the Augur PR and update here.

victorlin replied:
I've renamed to `--config-section` following [another change](https://github.com/nextstrain/WNV/pull/102#discussion_r2322728354) to accept a list of strings instead of dot-separated.

My thinking is that 'key' represents a single identifier in a key/value pair whereas 'section' refers more generally to a subset of the YAML.

joverlee521 reviewed (commented):
Thanks for going down this single config file route and revealing the changes needed for `augur subsample` to work well with this type of workflow! 

I left a smattering of comments, but the main thing I'd want to reconsider is the separation of nested build params vs top-level workflow params.

victorlin committed:
Consolidate all builds into one config / workflow run

The primary motivation is that having 'all-lineages' subsampling
configuration as default and using build-specific config overlays is a
fragile pattern, as it requires the build-specific configs to explicitly
overwrite "subsampling.regions" and "subsampling.force_include".

Even though build names are now provided under 'build_params', listing
them out explicitly via  'builds' is retained to prevent default builds
from leaking into workflow runs using config overlays, and provide an
easy way for CI to run just the all-lineages build.

Basic validation¹ has been added to ensure that 'builds_to_run' is
strictly a subset of available builds. There is more room for validation
such as checking that 'builds_to_run' is defined and as a list, but such
validation is lacking from the rest of the workflow so adding it now
seems out of scope.

¹ error message format taken from augur merge: https://github.com/nextstrain/augur/blob/e8b4d47fd9aae01eb538cc9225992606c7573149/augur/merge.py#L165-L171

joverlee521 reviewed (approved):
Thanks for all of the discussions around configs!

victorlin commented:
## Post-merge summary

All user-facing changes to the workflow have been summarized in the changelog:

> * Configuration resolved at run time is now written to a file under `results/run_configs`.
> * The phylogenetic workflow configuration now expects a new structure with most configuration defined under a build-specific key. See `phylogenetic/defaults/config.yaml` as an example. **This is a breaking change.**
> * The phylogenetic workflow now runs all Nextstrain-maintained builds by default. This can be adjusted by the `builds` configuration option. **This is a breaking change.**
> * Support for manual subsampling by editing the file `phylogenetic/rules/subsampling_manual.smk` is no longer supported. **This is a breaking change.**

There were many review comment threads, some with multiple discussion points. Each are linked with a short resolution note in the PR description. A summary of notable decisions which could manifest as standards for other repos to follow:

- We [decided](https://github.com/nextstrain/WNV/pull/102#discussion_r2320224474) to write config dumps to a folder instead of a constant file that gets updated. This slightly complicates downstream usage in `augur subsample` ([draft](https://github.com/nextstrain/WNV/commit/4d7364a0fcbfeffe0fa1e72c80a0cb87f245acca)), but I think it's actually a good thing.
- We [decided](https://github.com/nextstrain/WNV/pull/102#discussion_r2319726936) to move almost all config keys to build-specific sections. This seemed necessary for future-proofing the workflow. **Some downsides which had not been noted earlier**:
    - This effectively removes per-rule default values. The only defaults that are applied to external builds are `strain_id_field` and `inputs`.
    - Since they are no longer defaults, config shared across builds is now duplicated.
- We [discussed](https://github.com/nextstrain/WNV/pull/102#discussion_r2320205454) how to structure the config. Some terminology was defined ("build-first", "rule-first", "rule-only"). We noted that which the approach to adopt is a workflow-specific decision, but **decided** to standardize on a pattern of two top-level keys for "build-first": (1) a list of build names (`builds` in this repo) and (2) an object to contain build-specific parameters keyed by name (`build_params` in this repo).
- For [augur subsample](https://github.com/nextstrain/WNV/pull/96), we [decided](https://github.com/nextstrain/WNV/pull/102#discussion_r2323061426) the interface to target a specific section of YAML config with `--config-section`.
